Source/Core/Core/AchievementManager.cpp

```diff
@@ -1016,694 +520,461 @@ void AchievementManager::FilereaderClose(void* file_handle)
-    const std::string response_str(http_response->begin(), http_response->end());
-    if (process_response(rc_response, response_str.c_str()) != RC_OK)
+
+  m_image_queue.EmplaceItem([this, badge, badge_type, function, callback_data] {
+    std::string name_to_fetch;
     {
-      ERROR_LOG_FMT(ACHIEVEMENTS, "Failed to process HTTP response. \nURL: {} \nresponse: {}",
-                    api_request.url, response_str);
-      return ResponseType::MALFORMED_OBJECT;
+      std::lock_guard lg{m_lock};
+      if (function(*this).empty())
+        return;
+      name_to_fetch = function(*this);
     }
-    if (rc_response->response.succeeded)
+    rc_api_fetch_image_request_t icon_request = {.image_name = name_to_fetch.c_str(),
+                                                 .image_type = badge_type};
+    Badge fetched_badge;
+    rc_api_request_t api_request;
+    Common::HttpRequest http_request;
+    if (rc_api_init_fetch_image_request(&api_request, &icon_request) != RC_OK)
     {
-      return ResponseType::SUCCESS;
+      ERROR_LOG_FMT(ACHIEVEMENTS, "Invalid request for image {}.", name_to_fetch);
+      return;
     }
-    else
+    auto http_response = http_request.Get(api_request.url);
+    if (http_response.has_value() && http_response->size() <= 0)
     {
-      Logout();
-      WARN_LOG_FMT(ACHIEVEMENTS, "Invalid RetroAchievements credentials; failed login.");
-      return ResponseType::INVALID_CREDENTIALS;
+      WARN_LOG_FMT(ACHIEVEMENTS, "RetroAchievements connection failed on image request.\n URL: {}",
+                   api_request.url);
+      rc_api_destroy_request(&api_request);
+      m_update_callback(callback_data);
+      return;
     }
-  }
-  else
-  {
-    WARN_LOG_FMT(ACHIEVEMENTS, "RetroAchievements connection failed. \nURL: {}", api_request.url);
-    return ResponseType::CONNECTION_FAILED;
-  }
-}
 
-AchievementManager::ResponseType
-AchievementManager::RequestImage(rc_api_fetch_image_request_t rc_request, Badge* rc_response)
-{
-  rc_api_request_t api_request;
-  Common::HttpRequest http_request;
-  if (rc_api_init_fetch_image_request(&api_request, &rc_request) != RC_OK)
-  {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Invalid request for image.");
-    return ResponseType::INVALID_REQUEST;
-  }
-  auto http_response = http_request.Get(api_request.url);
-  if (http_response.has_value() && http_response->size() > 0)
-  {
     rc_api_destroy_request(&api_request);
-    *rc_response = std::move(*http_response);
-    return ResponseType::SUCCESS;
-  }
-  else
-  {
-    WARN_LOG_FMT(ACHIEVEMENTS, "RetroAchievements connection failed on image request.\n URL: {}",
-                 api_request.url);
-    rc_api_destroy_request(&api_request);
-    return ResponseType::CONNECTION_FAILED;
-  }
+    fetched_badge = std::move(*http_response);
+
+    INFO_LOG_FMT(ACHIEVEMENTS, "Successfully downloaded badge id {}.", name_to_fetch);
+    std::lock_guard lg{m_lock};
+    if (function(*this).empty() || name_to_fetch != function(*this))
```

Comment:
This logic seems odd to me.  You just finished downloading the badge and then you check if it is outdated?

It seems odd because why now?  What happens if after `badge->name` gets set the data is outdated?